risc0/circuit/rv32im/src/prove/emu/rv32im.rs

```diff
@@ -0,0 +1,622 @@
+    insn(InsnKind::ORI, InsnCategory::Compute, 0x13, 0x6, -1, 2),
+    insn(InsnKind::ANDI, InsnCategory::Compute, 0x13, 0x7, -1, 2),
+    insn(InsnKind::SLLI, InsnCategory::Compute, 0x13, 0x1, 0x00, 1),
+    insn(InsnKind::SRLI, InsnCategory::Compute, 0x13, 0x5, 0x00, 2),
+    insn(InsnKind::SRAI, InsnCategory::Compute, 0x13, 0x5, 0x20, 2),
+    insn(InsnKind::SLTI, InsnCategory::Compute, 0x13, 0x2, -1, 1),
+    insn(InsnKind::SLTIU, InsnCategory::Compute, 0x13, 0x3, -1, 1),
+    insn(InsnKind::BEQ, InsnCategory::Compute, 0x63, 0x0, -1, 1),
+    insn(InsnKind::BNE, InsnCategory::Compute, 0x63, 0x1, -1, 1),
+    insn(InsnKind::BLT, InsnCategory::Compute, 0x63, 0x4, -1, 1),
+    insn(InsnKind::BGE, InsnCategory::Compute, 0x63, 0x5, -1, 1),
+    insn(InsnKind::BLTU, InsnCategory::Compute, 0x63, 0x6, -1, 1),
+    insn(InsnKind::BGEU, InsnCategory::Compute, 0x63, 0x7, -1, 1),
+    insn(InsnKind::JAL, InsnCategory::Compute, 0x6f, -1, -1, 1),
+    insn(InsnKind::JALR, InsnCategory::Compute, 0x67, 0x0, -1, 1),
+    insn(InsnKind::LUI, InsnCategory::Compute, 0x37, -1, -1, 1),
+    insn(InsnKind::AUIPC, InsnCategory::Compute, 0x17, -1, -1, 1),
+    insn(InsnKind::MUL, InsnCategory::Compute, 0x33, 0x0, 0x01, 1),
+    insn(InsnKind::MULH, InsnCategory::Compute, 0x33, 0x1, 0x01, 1),
+    insn(InsnKind::MULHSU, InsnCategory::Compute, 0x33, 0x2, 0x01, 1),
+    insn(InsnKind::MULHU, InsnCategory::Compute, 0x33, 0x3, 0x01, 1),
+    insn(InsnKind::DIV, InsnCategory::Compute, 0x33, 0x4, 0x01, 2),
+    insn(InsnKind::DIVU, InsnCategory::Compute, 0x33, 0x5, 0x01, 2),
+    insn(InsnKind::REM, InsnCategory::Compute, 0x33, 0x6, 0x01, 2),
+    insn(InsnKind::REMU, InsnCategory::Compute, 0x33, 0x7, 0x01, 2),
+    insn(InsnKind::LB, InsnCategory::Load, 0x03, 0x0, -1, 1),
+    insn(InsnKind::LH, InsnCategory::Load, 0x03, 0x1, -1, 1),
+    insn(InsnKind::LW, InsnCategory::Load, 0x03, 0x2, -1, 1),
+    insn(InsnKind::LBU, InsnCategory::Load, 0x03, 0x4, -1, 1),
+    insn(InsnKind::LHU, InsnCategory::Load, 0x03, 0x5, -1, 1),
+    insn(InsnKind::SB, InsnCategory::Store, 0x23, 0x0, -1, 1),
+    insn(InsnKind::SH, InsnCategory::Store, 0x23, 0x1, -1, 1),
+    insn(InsnKind::SW, InsnCategory::Store, 0x23, 0x2, -1, 1),
+    insn(InsnKind::EANY, InsnCategory::System, 0x73, 0x0, 0x00, 1),
+    insn(InsnKind::MRET, InsnCategory::System, 0x73, 0x0, 0x18, 1),
+];
+
+// RISC-V instruction are determined by 3 parts:
+// - Opcode: 7 bits
+// - Func3: 3 bits
+// - Func7: 7 bits
+// In many cases, func7 and/or func3 is ignored.  A stardard trick is to decode
+// via a table, but a 17 bit lookup table destroys L1 cache.  Luckily for us,
+// in practice the low 2 bits of opcode are always 11, so we can drop them, and
+// also func7 is always either 0, 1, 0x20 or don't care, so we can reduce func7
+// to 2 bits, which gets us to 10 bits, which is only 1k.
+struct FastDecodeTable {
+    table: FastInstructionTable,
+}
+
+impl FastDecodeTable {
```

Comment:
@jbruestle this came from C++ code

Thing is, if this wasn't working, lots of other things would be broken. I also think a fuzzer is probably the right tool as a test for this, it'll be hard to make unit tests be complete I think

Reply:
Basically I think it's fine because we have the RISC-V test suite that passes, but we know that it's also not complete (we've found gaps in the past).

Reply:
Yup that checks out to me.